### PR TITLE
infoschema: tend to cache the old key in caching strategy for infoschema v2

### LINKED ISSUE #54500: Unrelated DDL change cause cache miss and reload in infoschema v2
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

As the title says, that kind of behavior is not expected. 

For example, if I have a table t1
I do some unrelated DDL operations
Visiting t1 should still always hit cache

### 2. What did you expect to see? (Required)

create table t1 (id int);
select * from t1;   // hit infoschema cache
// evit the schema cache of t1
create table t2 (id int);
select * from t1; 
select * from t1;   //  hit infoschema cache because DDL is on t2, unrelated to t1
create table t3 (id int);
select * from t1;   // hit infoschema cache because DDL is on t3, unrelated to t1
create table t4 (id int);
select * from t1;;  // hit infoschema cache because DDL is on t4, unrelated to t1

### 3. What did you see instead (Required)

After the original t1 cache  been evicted, the following DDL operation would cause schema cache miss everytime.

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #54500 
ref #50959 

Problem Summary:

### What changed and how does it work?

The bug is caused by the old caching strategy.
When we check the schema version from the btree, we get the old cache key.
When we update the cache, we use the latest schema version as the cache key.
It means when the old cache key is evicted, we will always cache miss and the refill does not help.

In this commit, I change the caching strategy tending to cache the old schema version.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by GMHDBJD — APPROVED** [review]
LGTM

File: pkg/infoschema/infoschema_v2.go [hunk 1]
```diff
@@ -444,6 +440,17 @@ func isSpecialDB(dbName string) bool {
 		dbName == util.MetricSchemaName.L
 }
 
+// EvictTable is exported for testing only.
+func (is *infoschemaV2) EvictTable(schema, tbl string) {
```
Comment (D3Hunter) [inline pkg/infoschema/infoschema_v2.go:444]: maybe implement it in test file of the same package, see https://github.com/pingcap/tidb/blob/ef0a3eb8af93da86fcf56443e58f0363584443e0/pkg/disttask/framework/scheduler/main_test.go#L25-L32

File: pkg/infoschema/test/infoschemav2test/v2_test.go [hunk 2]
```diff
@@ -263,3 +263,36 @@ func TestTiDBSchemaCacheSizeVariable(t *testing.T) {
...
+	is := dom.InfoSchema()
+	ok, v2 := infoschema.IsV2(is)
+	require.True(t, ok)
+	v2.EvictTable("test", "t1")
+
+	tk.MustExec("create table t2 (id int)")
+
+	// DDL on t2 should not cause reload or cache miss on t1
+	// before test, check failpoint works
+	failpoint.Enable("github.com/pingcap/tidb/pkg/infoschema/mockLoadTableInfoError", `return(true)`)
```
Comment (D3Hunter) [inline pkg/infoschema/test/infoschemav2test/v2_test.go:286]: use `testfailpoing.Enable/Disable`

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/infoschema/infoschema_v2.go (hunk 1 above)
Comment (tiancaiamao) [reply pkg/infoschema/infoschema_v2.go:444]: Some other parts may also need this functionality and it's in some other packages.
https://github.com/pingcap/tidb/pull/54479/files#diff-cabf59f012600c8b513d4eb03385f48e3fdd9b8ac3f13c175133d40c9354722bR279

**Review by D3Hunter — APPROVED** [review]

**Comment by tiancaiamao** [comment]
Need to wait this PR to merge first https://github.com/pingcap/tidb/pull/54282
The mysql-test has been merged, and that one is merging, so in the meantime the CI for master branch would fail temporarily

**Pull request merged**